### DISCUSSION TITLE
Recycling Lpuart pins

### DISCUSSION THREAD

**Discussion by isikkema** [mciantyre/teensy4-rs#135] [Q&A] [answered]
For reasons I can't explain, I'm trying to use a pin for `Lpuart` and then later as an `AnalogInput`.

I've tried this:
```rust
let mut lpuart2: board::Lpuart2 = board::lpuart(lpuart2, pins.p14, pins.p15, UART_BAUD);
/*
use lpuart2...
*/

// Pull pins out of Lpuart
lpuart2.reset();
let (_, lpuart::Pins { tx: p14, .. }) = lpuart2.release();

let mut analog_input = AnalogInput::new(p14);

log::info!("Pin 14 ADC: {}", adc1.read_blocking(&mut analog_input)); // Always reads a HIGH value (1022)
```

This compiles, but it only reads high values like 1022 from now on.

If I never use it for `Lpuart`, then it works as expected:
```rust
// let mut lpuart2: board::Lpuart2 = board::lpuart(lpuart2, pins.p14, pins.p15, UART_BAUD);
/*
use lpuart2...
*/

// Pull pins out of Lpuart
// lpuart2.reset();
// let (_, lpuart::Pins { tx: p14, .. }) = lpuart2.release();

let p14 = pins.p14;
let mut analog_input = AnalogInput::new(p14);

log::info!("Pin 14 ADC: {}", adc1.read_blocking(&mut analog_input)); // Reads the expected value
```

Is there a way to reset the pin to its pre-lpuart state?

**Comment by mciantyre** [comment, marked as answer]
I added `NEW` blocks into the troublesome example. Could you give something like this a try?

```rust
let mut lpuart2: board::Lpuart2 = board::lpuart(lpuart2, pins.p14, pins.p15, UART_BAUD);
/*
use lpuart2...
*/

// Pull pins out of Lpuart
lpuart2.reset();

//
// NEW: make p14 mut.
//
let (_, lpuart::Pins { tx: mut p14, .. }) = lpuart2.release();

//
// NEW: set the pin back to its GPIO alternate.
//
use teensy4_bsp::hal::iomuxc;
iomuxc::alternate(&mut p.p14, 5);

let mut analog_input = AnalogInput::new(p14);

log::info!("Pin 14 ADC: {}", adc1.read_blocking(&mut analog_input)); // Always reads a HIGH value (1022)
```

From what I can see in the code, there may be a missing configuration down in imxrt-iomuxc. The `adc::prepare()` call (a detail of `AnalogInput::new`) should set the pad alternate to a GPIO, but it's not doing that. By calling `iomuxc::alternate(..., 5)`, we're setting the pad back to a GPIO in our higher-level code.

If this works, we'll fix imxrt-iomuxc so that we won't need these `alternate` calls.

**Reply by isikkema** [reply]
This works. Thank you!

**Reply by isikkema** [reply]
I had previously looked in the code myself, and I saw things like alternate, daisy, sion, etc. that I wasn't able to find more information on in this context. Admittedly, I didn't look for TOO long.

Could you point me towards any documentation for these sorts of things?

**Reply by mciantyre** [reply]
Thanks for confirming! https://github.com/imxrt-rs/imxrt-iomuxc/issues/36 tracks a fix in imxrt-iomuxc.

Chapter 11 in the iMXRT1060 reference manual describes IOMUXC, the peripheral for pin muxing. It should qualify what SION, alternate values, daisy registers are. We could also summarize this in imxrt-iomuxc documentation.

If you're interested in diving in, [here](https://imxrt-rs.github.io/book/external_docs.html) is how we recommend acquiring reference manuals. <sub>(NXP requires a login to access reference manuals. But if you only care about one chip, [there's other ways](https://www.pjrc.com/teensy/datasheets.html) to find reference manuals...)</sub>
